CONTRIBUTING.md

```diff
@@ -0,0 +1,276 @@
+
+These should not be used in methods in a new class, since the whole class is new.
+
+The current in-development version can be found in `nextcord/__init__.py` at `__version__`.
+
+#### Versioning Example
+
+```python
+class Client:
+    """A client to interface with Discord.
+
+    This client is very barebones.
+
+    .. versionadded:: 2.0
+    """
+
+    def start(self, token: Optional[str] = None) -> None:
+        """Start the client.
+
+        .. versionadded:: 2.1
+
+        .. versionchanged:: 2.2
+
+            Added the `token` parameter.
+        """
+```
+
+### Operations
+
+We use [`.. container:: operations`](https://docutils.sourceforge.io/docs/ref/rst/directives.html#container) to describe special methods. This should be below the summary, and above the attributes.
+
+#### Operations Example
+
+```python
+class Guild(Hashable):
+    """Represents a Discord guild.
+
+    This is referred to as a "server" in the official Discord UI.
+
+    .. container:: operations
+
+        .. describe:: x == y
+
+            Checks if two guilds are equal.
+
+        .. describe:: x != y
+
+            Checks if two guilds are not equal.
+
+        .. describe:: hash(x)
+
+            Returns the guild's hash.
+
+        .. describe:: str(x)
+
+            Returns the guild's name.
+```
+
+### Class, Attribute, Method, ... Links
+
+We use [Roles](https://www.sphinx-doc.org/en/master/usage/restructuredtext/roles.html#roles) to link to other parts of documentation.
+
+`typing` is not linked to shorten the docstrings, and are simply described like this: ``List[:class:`str`]``.
+
+#### Links Example
+
+```python
+    @property
+    def members(self) -> List[Member]:
+        """List[:class:`Member`]: Returns all the members with this role."""
+
+        ,,,
+```
+
+### Coroutines
+
+Coroutines are denoted with `|coro|` on the first line, and `|maybecoro|` if it can be a coroutine or not (overriden methods).
+
+#### Coroutines Example
+
+```python
+async def sleep_until(...) -> Optional[T]:
+    """|coro|
+
+    Sleep until a specified time.
+
+    ...
+    """
+
+    ...
+```
+
+## Typed Payloads
+
+We use [TypedDict](https://docs.python.org/3/library/typing.html#typing.TypedDict) to type payloads. This is a dictionary with a specific set of keys, where the values have a specific type.
+
+These are stored in `nextcord/types`, where the file name relates to the `nextcord/` filename.
+
+These are Python representations of Discord payloads found in the [Discord API docs](https://discord.dev). [`typing_extensions.NotRequired`](https://peps.python.org/pep-0655/) is used for keys which are omitted, represented by `name?` in the Discord docs. [`typing.Optional`](https://docs.python.org/3/library/typing.html#typing.Optional) is used for keys which may be `null` (`None`), represented by `?type` in the Discord docs.
```

Comment:
For `NotRequired`, shouldn't you link to the Python Docs?

Reply:
typing_extensions is not in the docs

Reply:
True